### PR TITLE
Shipping Calculator Changes

### LINKED ISSUE #397: Await all promises in the same time in method getEligibleShippingMethods in ShippingCalculator
I have some plugins using ShippingCalculator that are promises. Would be nice to all promises were loaded at the same time. Currently, the method `getEligibleShippingMethods` awaits the response from the `shippingMethods`  individually.

https://github.com/vendure-ecommerce/vendure/blob/b2f3f082ade6e95374b997651ac517cb2c1a1890/packages/core/src/service/helpers/shipping-calculator/shipping-calculator.ts#L17-L32

A solution would be using [Promises.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all).

### PR DESCRIPTION
**Pull request by jonyw4**
Closes #397 and #398

### CODE REVIEW

File: packages/core/src/service/helpers/shipping-calculator/shipping-calculator.ts [hunk 1]
```diff
@@ -14,21 +23,32 @@ export class ShippingCalculator {
...
-                    eligibleMethods.push({ method, result });
-                }
+
+        const checkEligibilityPromises = shippingMethods.map((method) =>
+            this.checkEligibilityByShippingMethod(order, method),
+        );
+        const eligibleMethods = await Promise.all(checkEligibilityPromises);
+
+        // Workaround to remove undefined from type: https://stackoverflow.com/a/51577579
+        const filteredEligibleMethods = eligibleMethods.filter(
```
Comment (michaelbromley) [inline packages/core/src/service/helpers/shipping-calculator/shipping-calculator.ts]: You can use the `notNullOrUndefined` util function here (https://github.com/vendure-ecommerce/vendure/blob/master/packages/common/src/shared-utils.ts#L5)

You can import it from "@vendure/common/lib/shared-utils"

File: packages/core/src/service/helpers/shipping-calculator/shipping-calculator.ts [hunk 2]
```diff
@@ -6,6 +6,15 @@ import { Order } from '../../../entity/order/order.entity';
 import { ShippingMethod } from '../../../entity/shipping-method/shipping-method.entity';
 import { ShippingMethodService } from '../../services/shipping-method.service';
 
+type CheckEligibilityByShippingMethodResponse =
```
Comment (michaelbromley) [inline packages/core/src/service/helpers/shipping-calculator/shipping-calculator.ts]: I would advise to just define a single type `CheckEligibilityByShippingMethodResponse ` without the `undefined` union, and then in define like `checkEligibilityByShippingMethod(): Promise<CheckEligibilityByShippingMethodResponse | undefined>`. This requires one less type definition and IMO is simpler to understand.

File: packages/core/src/service/helpers/shipping-calculator/shipping-calculator.ts (hunk 1 above)
Comment (jonyw4) [reply packages/core/src/service/helpers/shipping-calculator/shipping-calculator.ts]: Great! When I wrote this part using filter the type definitions didn't get right. Good to know that have a function to deal with this 😁

File: packages/core/src/service/helpers/shipping-calculator/shipping-calculator.ts (hunk 2 above)
Comment (jonyw4) [reply packages/core/src/service/helpers/shipping-calculator/shipping-calculator.ts]: Nice! Good point. I will update

**Comment by jonyw4** [comment]
> See a couple of comments. Also, could you rebase onto the latest `next` branch because I fixed the causes of the failing tests, so then we know whether these changes cause any tests to fail.

Done! 😬

**Pull request merged**